### ISSUE TITLE
photon is completely broken

### ISSUE THREAD

**Issue by ssahani** [systemd/mkosi#664]
```
`commit 0f167a2205070096e182222a84ddc954a6c23997
Author: Daan De Meyer <daan.j.demeyer@gmail.com>
Date:   Wed Dec 2 19:57:48 2020 +0000

```

@DaanDeMeyer your commit broke photon

/cc @smaliakkal

**Comment by behrmann** [comment]
That commit only removed CI for Photon. Could you describe what the problem is, on which version of mkosi it appears, how you call mkosi (i.e. provide a minimal example to reproduce this issue) and a version where this last worked for you? If you could bisect the specific commit that broke you, that would be tremendously helpful, but the other infos would be a great start.

**Comment by ssahani** [comment]
See this a8a14b8bfe39d91f933062667877bfc0de686541.  Have you tested even before you make. changes ?

**Comment by DaanDeMeyer** [comment]
Unfortunately, half the features didn't even work when I tried (for a very long time) to add Photon to the CI. I even pinged you for help to get Photon working in CI but you never responded so we didn't add it to CI so we can't guarantee it keeps working. If you do the work of maintaining the photon implementation and put in the effort to add it to the CI (full feature set including QEMU and nspawn boot), I'll happily make sure it stays working.

But you have to understand that we just can't make progress if every contributor has to manually test every niche distro supported by mkosi without getting any support from the person that originally added it.

TLDR; Make PRs to fix all the issues with Photon and add Photon to mkosi CI and we'll make sure it doesn't break again.

**Comment by behrmann** [comment]
Back then there was no CI for Photon and CI for Photon was only added in f06b436899ea4502634e868e64ab53ee5407db08 (by @DaanDeMeyer), but removed again in 0f167a2205070096e182222a84ddc954a6c23997, since it was too much of a maintenance burden and generated images already didn't boot (if I remember correctly). Neither is the fault of @DaanDeMeyer. 

If you provide a reproducer, we can still have a look what the issue is.

**Comment by ssahani** [comment]
> Back then there was no CI for Photon and CI for Photon was only added in [f06b436](https://github.com/systemd/mkosi/commit/f06b436899ea4502634e868e64ab53ee5407db08) (by @DaanDeMeyer), but removed again in [0f167a2](https://github.com/systemd/mkosi/commit/0f167a2205070096e182222a84ddc954a6c23997), since it was too much of a maintenance burden and generated images already didn't boot (if I remember correctly). Neither is the fault of @DaanDeMeyer.
> 
> If you provide a reproducer, we can still have a look what the issue is.

just try to generate the photon image, it will break. Checkout the  commit we submitted that will work ```git checkout a8a14b8bfe39d91f933062667877bfc0de686541```

**Comment by behrmann** [comment]
> > If you provide a reproducer, we can still have a look what the issue is.
> 
> just try to generate the photon image, it will break. Checkout the commit we submitted that will work `git checkout a8a14b8bfe39d91f933062667877bfc0de686541`

Ok, I'll try it out once I have some downtime, but I will have to either spin up a Photon VM or build tdnf myself to have a look at this. The latter route is more promising, since without adding Photon back to CI, there is a non-negligible probability that it will break again in the future. 

What would help is to have a mkosi invocation that breaks and to see a stacktrace before anybody gets to work. :)

**Comment by ssahani** [comment]
> Unfortunately, half the features didn't even work wh di en I tried (for a very long time) to add Photon to the CI. I even pinged you for help to get Photon working in CI but you never responded so we didn't add it to CI so we can't guarantee it keeps working. If you do the work of maintaining the photon implementation and put in the effort to add it to the CI (full feature set including QEMU and nspawn boot), I'll happily make sure it stays working.
> 
> But you have to understand that we just can't make progress if every contributor has to manually test every niche distro supported by mkosi without getting any support from the person that originally added it.
> 
> TLDR; Make PRs to fix all the issues with Photon and add Photon to mkosi CI and we'll make sure it doesn't break again.

I don't understand you TLDRl. Can you explain why you merged without testing.

**Comment by ssahani** [comment]
> > > If you provide a reproducer, we can still have a look what the issue is.
> > 
> > 
> > just try to generate the photon image, it will break. Checkout the commit we submitted that will work `git checkout a8a14b8bfe39d91f933062667877bfc0de686541`
> 
> Ok, I'll try it out once I have some downtime, but I will have to either spin up a Photon VM or build tdnf myself to have a look at this. The latter route is more promising, since without adding Photon back to CI, there is a non-negligible probability that it will break again in the future.
> 
> What would help is to have a mkosi invocation that breaks and to see a stacktrace before anybody gets to work. :)

You run that and you will get the backtrace.

**Comment by mrc0mmand** [comment]
> > Unfortunately, half the features didn't even work wh di en I tried (for a very long time) to add Photon to the CI. I even pinged you for help to get Photon working in CI but you never responded so we didn't add it to CI so we can't guarantee it keeps working. If you do the work of maintaining the photon implementation and put in the effort to add it to the CI (full feature set including QEMU and nspawn boot), I'll happily make sure it stays working.
> > But you have to understand that we just can't make progress if every contributor has to manually test every niche distro supported by mkosi without getting any support from the person that originally added it.
> > TLDR; Make PRs to fix all the issues with Photon and add Photon to mkosi CI and we'll make sure it doesn't break again.
> 
> I don't understand you TLDRl. Can you explain why you merged without testing.

See the https://github.com/systemd/mkosi/pull/542/commits/0f167a2205070096e182222a84ddc954a6c23997 commit message for the reasoning:

```
The overhead of supporting these in CI is too big compared to the
benefit of maintaining them in CI. We can add back support for these
distros once a dedicated maintainer shows up willing to do the work.
```

These CIs were dropped, because there's no one who would volunteer to keep them alive and we simply don't have enough workforce to do it, as the current set already consumes an insignificant amount of time (see the recent Arch & glibc issue). So, the tl;dr is - if you want Photon in the current set of active CIs, you'd need to maintain it to keep it *green*.

**Comment by ssahani** [comment]
> > > Unfortunately, half the features didn't even work wh di en I tried (for a very long time) to add Photon to the CI. I even pinged you for help to get Photon working in CI but you never responded so we didn't add it to CI so we can't guarantee it keeps working. If you do the work of maintaining the photon implementation and put in the effort to add it to the CI (full feature set including QEMU and nspawn boot), I'll happily make sure it stays working.
> > > But you have to understand that we just can't make progress if every contributor has to manually test every niche distro supported by mkosi without getting any support from the person that originally added it.
> > > TLDR; Make PRs to fix all the issues with Photon and add Photon to mkosi CI and we'll make sure it doesn't break again.
> > 
> > 
> > I don't understand you TLDRl. Can you explain why you merged without testing.
> 
> See the [0f167a2](https://github.com/systemd/mkosi/commit/0f167a2205070096e182222a84ddc954a6c23997) commit message for the reasoning:
> 
> ```
> The overhead of supporting these in CI is too big compared to the
> benefit of maintaining them in CI. We can add back support for these
> distros once a dedicated maintainer shows up willing to do the work.
> ```
> 
> These CIs were dropped, because there's no one who would volunteer to keep them alive and we simply don't have enough workforce to do it, as the current set already consumes an insignificant amount of time (see the recent Arch & glibc issue). So, the tl;dr is - if you want Photon in the current set of active CIs, you'd need to maintain it to keep it _green_.

That does not mean You break the the workflow.  And why you added that change

**Comment by ssahani** [comment]
It's not about maintaining th eCI . You actually broke the entire component.

**Comment by mrc0mmand** [comment]
@DaanDeMeyer didn't break anything, the CIs for the distros mentioned in the commit were already failing, i.e. they were already most likely broken. The commit you keep referencing didn't remove anything from the actual codebase, just dropped the aforementioned distros from the CI matrix.

**Comment by ssahani** [comment]
> @DaanDeMeyer didn't break anything, the CIs for the distros mentioned in the commit were already failing, i.e. they were already most likely broken. The commit you keep referencing didn't remove anything from the actual codebase, just dropped the aforementioned distros from the CI matrix.

So you saying the patch never worked.  You break and never admit

**Comment by DaanDeMeyer** [comment]
Of course I might have broken photon. tdnf is not available in the AUR and I'm not going to go throught the effort of compiling it from source to test every change I make with a niche distro such as photon. Our policy is that if it's not in CI, we don't guarantee a distro stays working as changes are made to mkosi. That's why I said, if you want to make sure we don't break Photon when making changes, you should go do the effort to add it to the CI. If it's in the CI, it will let me know when a change I make breaks Photon. Until we have it in CI, every new commit could potentially break Photon. Since you're very likely the only user of mkosi that's using Photon, It's on you to make sure it keeps working as we push new commits and provide us with detailed bug reports including a stacktrace when something fails. We don't use Photon and we don't have tdnf installed so we can't build and test Photon when we make changes.

I had the same issue with systemd, every so often they broke running the unit tests in mkosi so I added mkosi to systemd's CI and since then @mrc0mmand has been very helpful in making sure they keep passing. If you do the same effort and make sure Photon works and keeps working in mkosi's CI, then we notice when we break Photon's build and we can fix it immediately.

**Comment by ssahani** [comment]
> Of course I might have broken photon. tdnf is not available in the AUR and I'm not going to go throught the effort of compiling it from source to test every change I make with a niche distro such as photon. Our policy is that if it's not in CI, we don't guarantee a distro stays working as changes are made to mkosi. That's why I said, if you want to make sure we don't break Photon when making changes, you should go do the effort to add it to the CI. If it's in the CI, it will let me know when a change I make breaks Photon. Until we have it in CI, every new commit could potentially break Photon. Since you're very likely the only user of mkosi that's using Photon, It's on you to make sure it keeps working as we push new commits and provide us with detailed bug reports including a stacktrace when something fails. We don't use Photon and we don't have tdnf installed so we can't build and test Photon when we make changes.
> 
> I had the same issue with systemd, every so often they broke running the unit tests in mkosi so I added mkosi to systemd's CI and since then @mrc0mmand has been very helpful in making sure they keep passing. If you do the same effort and make sure Photon works and keeps working in mkosi's CI, then we notice when we break Photon's build and we can fix it immediately.

you make effort to test rather giving lecture me to do effort . Why you broke first place ?

**Comment by ssahani** [comment]
It does not matter what you said. You need to test and test.

**Comment by ssahani** [comment]
your commits break and you talk abt guarantee.  Your ranting does not solve anything

**Comment by lucasdemarchi** [comment]
@ssahani one comment here every few minutes ain't helping nobody. As other people stated above, it's hard to reproduce it due to photon dependencies. If you have a reproducer, please paste the output @behrmann asked. This would allow people to figure out faster why it's not working.

**Comment by ssahani** [comment]
> @ssahani one comment here every few minutes ain't helping nobody. As other people stated above, it's hard to reproduce it due to photon dependencies. If you have a reproducer, please paste the output @behrmann asked. This would allow people to figure out faster why it's not working.

Just ignore. don's read

**Comment by johannbg** [comment]
> Of course I might have broken photon. tdnf is not available in the AUR and I'm not going to go throught the effort of compiling it from source to test every change I make with a niche distro such as photon. 

@DaanDeMeyer I noticed that in the comment that @ssahani references to you also drop OpenSuse which is one of the parent distribution ( read as not a niche distro or a derivative like Ubuntu or RHEL/CentOS ) which makes little to no sense to do, What's the justification of removing it? 

What exactly is that maintenance burden? 
Why keep those distributions literally hard coded if the intent is to never test them thus support them?

**Comment by DaanDeMeyer** [comment]
@johannbg OpenSuse was dropped because I couldn't get their package manager to build from source on Ubuntu after we had to start building librpm from source in order to keep dnf working in CI which broke the bundled version of zypper in Focal. If someone goes through the effort to get zypper working again, we can consider adding Suse back to CI.

> Why keep those distributions literally hard coded if the intent is to never test them thus support them?

Support was already there before I started maintaining the project. i'm not going to remove the support alltogether just because I can't verify if it's working or not. For all I know, people might be successfully using that distro with mkosi.

**Comment by johannbg** [comment]
> @johannbg OpenSuse was dropped because I couldn't get their package manager to build from source on Ubuntu after we had to start building librpm from source in order to keep dnf working in CI which broke the bundled version of zypper in Focal. If someone goes through the effort to get zypper working again, we can consider adding Suse back to CI.

Why dont you revert *your* changes that forced you to start building librpm from source to keep DNF working if that's what's breaking this for all the other distro's? It makes absolutely no sense implementing it when it breaks so many things for other distro's. 

> 
> > Why keep those distributions literally hard coded if the intent is to never test them thus support them?
> 
> Support was already there before I started maintaining the project. i'm not going to remove the support alltogether just because I can't verify if it's working or not. For all I know, people might be successfully using that distro with mkosi.

You do realize it's your own changes that broke this for other distribution right? So it goes without saying that more of your own changes can break this for other distributions as well thus it makes no sense keep all those distro's if in the end, the intent is to only test and support the RHEL ecosystem, which everyone is abandoning for (Open)Suse and Ubuntu/Debian, Rocky and something called Alma apparently as well ( which you would probably end up having to add support for in the long run ).

If this was Arch would you have allowed a RH ecosystem component to break the Arch support in mkosi and you would just end up removing Arch from the CI's ?

**Comment by behrmann** [comment]
Please, we all want the same things here and I don't think the current discussion is bringing anyone any closer :)

I for one don't find it productive to rant here and assign blame. Most of the CI coverage only arrived with @DaanDeMeyer and he later removed some distros from it again, because he probably didn't have the time to maintain it any further, after adding some actual functional tests to CI, i.e. checking that something actually boots after building the image. Nobody stepped up to maintain the CI for those distros and bug reports only come in occasionally, like now.

Before the broader CI was added and after it was removed for maintenance reasons, we could not know and we cannot know now whether something doesn't work if people don't open issues. Without bisecting the issue (which I hope to do next week), we don't know what broke it, but the probability is high that it was some change by @DaanDeMeyer, since he is by far the most prolific dev here, or me, since I tend to look at everything coming through and I do might miss things. 

Nobody is perfect and there is no malice in any changes. It is good that this issue was opened, although I would have preferred a different tone, and it will be addressed when somebody gets around to doing the required work. 

@ me if some actionable info comes around, I'll mute this thread for me now.

**Comment by ssahani** [comment]
> Please, we all want the same things here and I don't think the current discussion is bringing anyone any closer :)
> 
> I for one don't find it productive to rant here and assign blame. Most of the CI coverage only arrived with @DaanDeMeyer and he later removed some distros from it again, because he probably didn't have the time to maintain it any further, after adding some actual functional tests to CI, i.e. checking that something actually boots after building the image. Nobody stepped up to maintain the CI for those distros and bug reports only come in occasionally, like now.
> 
> Before the broader CI was added and after it was removed for maintenance reasons, we could not know and we cannot know now whether something doesn't work if people don't open issues. Without bisecting the issue (which I hope to do next week), we don't know what broke it, but the probability is high that it was some change by @DaanDeMeyer, since he is by far the most prolific dev here, or me, since I tend to look at everything coming through and I do might miss things.
> 
> Nobody is perfect and there is no malice in any changes. It is good that this issue was opened, although I would  have preferred a different tone, and it will be addressed when somebody gets around to doing the required work.
> 
> @ me if some actionable info comes around, I'll mute this thread for me now.

Now this is a sensible reply .  Giving lectures and ranting about policy @DaanDeMeyer  who broke does not help.  We already have the fix we will open up PR and follow up ci.

**Comment by poettering** [comment]
I am very thankful for the work @DaanDeMeyer is doing, and I think it's entirely Ok and expected if support for OSes for which there is no CI will sometimes get broken, and I am very sure there's no ill-intent in any of that.

CI is the only reasonable way to make sure these things don't get broken. Hence, please step up, maintain a CI for photon, and we are good and things won't break accidentally anymore — but unless that happens you *must* expect brokenness every now and then. 

Hence, if you submit support for anothre distro you basically have two options:

1. provide a CI too, and thus be reasonably sure things won't break accidentally for your distro
2. do not provide a CI, but then accept that your distro will likely break sooner or later, but not due to malice or ill intent, but simply because noone can check it. In this mode, send a patch to fix it up once you notice, but please stay friendly. we are all volunteers here!

**Comment by ssahani** [comment]
> I am very thankful for the work @DaanDeMeyer is doing, and I think it's entirely Ok and expected if support for OSes for which there is no CI will sometimes get broken, and I am very sure there's no ill-intent in any of that.
> 
> CI is the only reasonable way to make sure these things don't get broken. Hence, please step up, maintain a CI for photon, and we are good and things won't break accidentally anymore — but unless that happens you _must_ expect brokenness every now and then.
> 
> Hence, if you submit support for anothre distro you basically have two options:
> 
> 1. provide a CI too, and thus be reasonably sure things won't break accidentally for your distro
> 2. do not provide a CI, but then accept that your distro will likely break sooner or later, but not due to malice or ill intent, but simply because noone can check it. In this mode, send a patch to fix it up once you notice, but please stay friendly. we are all volunteers here!

we trying to stay friendly. Ranting about policy and giving lectures how does that help.  You missing the point of tests before merging anything
